File: OS_codes/PRA_FIFO.java
/* import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class PRA_FIFO {

    public static int fifoPageFaults(String referenceString, int frameSize) {
        int pageFaults = 0;
        Queue<Character> frames = new LinkedList<>();

        for (char page : referenceString.toCharArray()) {
            if (!frames.contains(page)) {
                if (frames.size() < frameSize) {
                    frames.add(page);
                } else {
                    char removedPage = frames.poll();
                    frames.add(page);
                    System.out.println("Page " + removedPage + " removed. Page " + page + " added. Frames: " + frames);
                    pageFaults++;
                }
            } else {
                System.out.println("Page " + page + " is already in memory. Frames: " + frames);
            }
        }

        return pageFaults;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the reference string: ");
        String referenceString = scanner.nextLine();

        System.out.print("Enter the frame size: ");
        int frameSize = scanner.nextInt();

        int result = fifoPageFaults(referenceString, frameSize);
        System.out.println("Total number of page faults using FIFO: " + result);

        scanner.close();
    }
}
 */

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class PRA_FIFO {

    private static final int MAX_FRAMES = 3;
    private static final int MAX_REFERENCES = 10;

    private int[] referenceString = {0, 1, 2, 3, 2, 4, 3, 4, 5, 3};
    private int[] frames = new int[MAX_FRAMES];
    private Queue<Integer> frameQueue = new LinkedList<>();

    private int pageFaults = 0;

    private void initializeFrames() {
        Arrays.fill(frames, -1); // Initialize frames to -1 indicating empty
    }

    private boolean isPageInFrames(int page) {
        for (int frame : frames) {
            if (frame == page) {
                return true; // Page found in frames
            }
        }
        return false; // Page not found in frames
    }

    private void simulateFIFO() {
        initializeFrames();

        System.out.print("Reference String: ");
        for (int i = 0; i < MAX_REFERENCES; i++) {
            int page = referenceString[i];

            System.out.print(page + " ");

            if (!isPageInFrames(page)) {
                int oldestFrame = frameQueue.poll();
                frameQueue.add(page);
                frames[oldestFrame] = page;
                pageFaults++;

                System.out.print("(Page Fault) ");
            }

            System.out.print("Frames: ");
            for (int frame : frames) {
                if (frame == -1) {
                    System.out.print("  ");
                } else {
                    System.out.print(frame + " ");
                }
            }
            System.out.println();
        }

        System.out.println("Total Page Faults: " + pageFaults);
    }

    public static void main(String[] args) {
        System.out.println("FIFO Page Replacement Algorithm Simulation");

        PRA_FIFO fifoSimulation = new PRA_FIFO();
        fifoSimulation.simulateFIFO();
    }
}

File: OS_codes/PRA_FIFO.c
#include <stdio.h>
#include <stdlib.h>

int main()
{
    int rs[25], frame[10], count, n, f, pf;
    printf("\nEnter the length of reference string- ");
    scanf("%d", &n);
    printf("\nEnter the reference string- ");
    for (int i = 0; i < n; i++)
        scanf("%d", &rs[i]);

    printf("\nEnter number of frames- ");
    scanf("%d", &f);

    for (int i = 0; i < f; i++)
        frame[i] = -1;

    count = 0;
    pf = 0;

    for (int i = 0; i < n; i++)
    {
        int found = 0;

        for (int j = 0; j < f; j++)
        {
            if (frame[j] == rs[i])
            {
                found = 1;
                break;
            }
        }

        if (found == 0)
        {
            int flag = 0;
            for (int j = 0; j < f; j++)
            {
                if (frame[j] == -1)
                {
                    flag = 1;
                    frame[j] = rs[i];
                    break;
                }
            }

            if (flag == 0)
            {
                frame[count] = rs[i];
                count++;

                if (count == f)
                    count = 0;
            }

            pf++;
        }

        printf("\n%d ||", rs[i]);
        for (int j = 0; j < f; j++)
        {
            printf("\t%d", frame[j]);
        }
        if (found == 0)
            printf("\tpf. count- %d", pf);
    }
    int hit = n - pf;
    float hitr = 100*(float)hit/n;
    float miss = 100*(float)pf/n;
    printf("\nTotal page faults in fifo= %d\n", pf);
    printf("Total hit= %d\n", hit);
    printf("Hit ratio = %0.2f percentage\n", hitr);
    printf("Miss ratio = %0.2f percentage\n", miss);
    return 0;
}
